Proxy control plane rate limiter

khanova opened the pull request:
## Problem

Proxy might overload the control plane.

## Summary of changes

Implement rate limiter for proxy<->control plane connection.
Resolves https://github.com/neondatabase/neon/issues/5707

Used implementation ideas from https://github.com/conradludgate/squeeze/

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

vadim2404 commented on proxy/src/rate_limiter/limit_algorithm.rs:
```diff
@@ -0,0 +1,61 @@
+impl Default for RateLimiterConfig {
+    fn default() -> Self {
+        Self {
+            algorithm: RateLimitAlgorithm::None,
```
if you make this

```rust

#[derive(Default, Clone, Copy, Debug, clap::ValueEnum)]
pub enum RateLimitAlgorithm {
    #[default]
    None,
    Fixed,
    Aimd,
}
```

it will allow you to have 

```rust
#[derive(Default)]
pub struct RateLimiterConfig
```

w/o explicitly overridden method.

khanova replied:
Done.

vadim2404 commented on proxy/src/bin/proxy.rs:
```diff
@@ -60,7 +61,7 @@ struct ProxyCliArgs {
     /// path to TLS key for client postgres connections
     ///
-    /// tls-key and tls-cert are for backwards compatibility, we can put all certs in one dir
+    /// tls-key and tls-cert are for backwards compatibility, we can put/// /// all certs in one dir
```
Unintentional changes?

khanova replied:
Fixed.

conradludgate commented:
Let me give some backstory on the [pin-list semaphore idea](https://github.com/conradludgate/squeeze/blob/3b35ff4570ace5b411c7022a5165becf53260b3f/src/limiter.rs#L287-L344) - Although I see it's not included in this current PR (but the pin-list dependency is still there)

## why was it necessary

I needed a way to remove available permits at will - eg when the congestion control algorithm wants to decrease the available concurrency.

`tokio::sync::Semaphore` only allows adding additional permits. To remove permits, you need to await an acquire and then `forget` them. This will have a lagged effect as these acquire calls with be at the back of the queue.

## Solution

Write our own semaphore. Easier said than done. Inspired by tokio's implementation which uses a [Mutex<LinkedList>](https://docs.rs/tokio/latest/src/tokio/sync/batch_semaphore.rs.html#36) where the linked list is intrusive. Writing intrusive linked lists requires unsafe code which I would like to avoid. Thankfully, a friend of mine authored this crate called [pin-list](https://docs.rs/pin-list/latest/pin_list/) which has a safe abstraction for an efficient intrusive-linked list.

I took the code from tokio, and built it on top of pin-list and made a few notable changes:
1) Instead of a [single `permits` available coutner](https://docs.rs/tokio/latest/src/tokio/sync/batch_semaphore.rs.html#38), I switched to [2 counters](https://github.com/conradludgate/squeeze/blob/3b35ff4570ace5b411c7022a5165becf53260b3f/src/limiter.rs#L43-L46)
2) I removed some of the code that dealt with acquiring multiple permits as it added a bit of extra unnecessary complexity

khanova commented:
> Let me give some backstory on the [pin-list semaphore idea](https://github.com/conradludgate/squeeze/blob/3b35ff4570ace5b411c7022a5165becf53260b3f/src/limiter.rs#L287-L344) - Although I see it's not included in this current PR (but the pin-list dependency is still there)
> 
> ## why was it necessary
> I needed a way to remove available permits at will - eg when the congestion control algorithm wants to decrease the available concurrency.
> 
> `tokio::sync::Semaphore` only allows adding additional permits. To remove permits, you need to await an acquire and then `forget` them. This will have a lagged effect as these acquire calls with be at the back of the queue.
> 
> ## Solution
> Write our own semaphore. Easier said than done. Inspired by tokio's implementation which uses a [`Mutex<LinkedList>`](https://docs.rs/tokio/latest/src/tokio/sync/batch_semaphore.rs.html#36) where the linked list is intrusive. Writing intrusive linked lists requires unsafe code which I would like to avoid. Thankfully, a friend of mine authored this crate called [pin-list](https://docs.rs/pin-list/latest/pin_list/) which has a safe abstraction for an efficient intrusive-linked list.
> 
> I took the code from tokio, and built it on top of pin-list and made a few notable changes:
> 
> 1. Instead of a [single `permits` available coutner](https://docs.rs/tokio/latest/src/tokio/sync/batch_semaphore.rs.html#38), I switched to [2 counters](https://github.com/conradludgate/squeeze/blob/3b35ff4570ace5b411c7022a5165becf53260b3f/src/limiter.rs#L43-L46)
> 2. I removed some of the code that dealt with acquiring multiple permits as it added a bit of extra unnecessary complexity

I see, yes, implementing semaphore is a non-trivial task. 

Originally I wanted to reuse your implementation with `pin-list`, but it looks like the bot is not very happy about this dependency.

My assumption was that it should be very rare situation when the control plane is overloaded and there is still non-trivial amount of available permits. Than it is not a problem to `forget` permits on the release. But this assumption might be completely wrong.

What do you think about it?

hlinnaka commented on proxy/src/bin/proxy.rs:
```diff
@@ -92,6 +93,27 @@ struct ProxyCliArgs {
+    #[clap(value_enum, long, default_value_t = proxy::rate_limiter::RateLimitAlgorithm::Aimd)]
+    rate_limit_algorithm: proxy::rate_limiter::RateLimitAlgorithm,
+    /// Initial limit for dynamic rate limiter. Makes sense only if `rate_limit_algorithm` is *not* `None`.
+    #[clap(long, default_value_t = 10)]
+    initial_limit: usize,
```
10 isn't very much. How fast does the algorithm ramp up?

hlinnaka commented on proxy/src/bin/proxy.rs:
```diff
@@ -92,6 +93,27 @@ struct ProxyCliArgs {
+    /// Initial limit for dynamic rate limiter. Makes sense only if `rate_limit_algorithm` is *not* `None`.
+    #[clap(long, default_value_t = 10)]
+    initial_limit: usize,
+    /// Minimum limit for AIMD algorithm. Makes sense only if `rate_limit_algorithm` is `Aimd`.
+    #[clap(long, default_value_t = 1)]
+    aimd_min_limit: usize,
```
1 feels very little. Do we really want to throttle down that much?

khanova replied:
I set these numbers almost randomly :) Probably, it does make sense to ramp a bit faster.

hlinnaka commented on proxy/src/bin/proxy.rs:
```diff
@@ -92,6 +93,27 @@ struct ProxyCliArgs {
+    /// Minimum limit for AIMD algorithm. Makes sense only if `rate_limit_algorithm` is `Aimd`.
+    #[clap(long, default_value_t = 1)]
+    aimd_min_limit: usize,
+    /// Maximum limit for AIMD algorithm. Makes sense only if `rate_limit_algorithm` is `Aimd`.
+    #[clap(long, default_value_t = 1000)]
+    aimd_max_limit: usize,
```
Hmm, my initial reaction is that 1000 is not very much for a maximum. But I'm not sure. How many concurrent requests can one control plane instance handle? What about all the control plane instances altogether?

khanova replied:
Hm, from [grafana dashboard](https://neonprod.grafana.net/d/S3MPUUonz/neon-console?orgId=1&refresh=5m) I see that ~500 goroutines is about 100% of CPU utilization for the console.
But there is also [grafana dashboard](https://neonprod.grafana.net/d/b74c1a8d-aa7e-48e6-a951-c92a2dde0996/neon-control-plane?orgId=1&refresh=5m) that shows that ~50 goroutines is about 1.5% of CPU utilization for the control plane.

I am a bit confused about that.

Probably @ololobus has a better understanding what is the right data.

ololobus replied:
> But there is also [grafana dashboard](https://neonprod.grafana.net/d/b74c1a8d-aa7e-48e6-a951-c92a2dde0996/neon-control-plane?orgId=1&refresh=5m) that shows that ~50 goroutines is about 1.5% of CPU utilization for the control plane.

> I am a bit confused about that.

These new control-planes are mostly idle, it's better to look on console dashboard for now

ololobus replied:
> Hmm, my initial reaction is that 1000 is not very much for a maximum. But I'm not sure. How many concurrent requests can one control plane instance handle? What about all the control plane instances altogether?

It's hard to answer right away, because we didn't do benchmarks / cap tests in close to prod env, but on my M2 laptop with Postgres in Docker one cplane instance gives
- 10k RPS for ProxyGetRoleSecret
- 200-300 RPS for ProxyWakeCompute

Both done from 200 clients with the same endpoint_id

hlinnaka reviewed (commented):
Some prometheus metrics to monitor the rate limiting would be nice.

If I understand correctly, this is a global rate limit on the number of requests to the control plane. Does it have any "fairness" built into it? If one user sends a lot of requests, can it saturate the limiter easily, effectively causing an outage for everyone else?

In production, we currently run three console/control plane instances behind a load balancer. If one of them is overloaded for some reason and fails all requests, but others are working correctly, how does the rate limiting algorithm behave? In the future, we will also have separate control plane instances in each region.

More tests would be nice. There are unit tests for the algorithm, but I'd also like to see some python tests, testing the throttling in the real proxy.

hlinnaka commented:
> If I understand correctly, this is a global rate limit on the number of requests to the control plane. Does it have any "fairness" built into it? If one user sends a lot of requests, can it saturate the limiter easily, effectively causing an outage for everyone else?

I see that https://github.com/neondatabase/neon/pull/5799 addresses that, with a per-endpoint lock. When both of these PRs are merged, I presume we will acquire the per-endpoint lock first, and the global limiter permit after that. That seems OK. Per IP address limiting would be nice too, to avoid DoSsing the control plane with 'get_auth_info' requests or saturating this rate limiter, but that's a different story.

conradludgate commented on proxy/src/bin/proxy.rs:
```diff
@@ -95,6 +96,33 @@ struct ProxyCliArgs {
+    /// Initial limit for dynamic rate limiter. Makes sense only if `rate_limit_algorithm` is *not* `None`.
+    #[clap(long, default_value_t = 100)]
+    initial_limit: usize,
+    /// Minimum limit for AIMD algorithm. Makes sense only if `rate_limit_algorithm` is `Aimd`.
+    #[clap(long, default_value_t = 1)]
+    aimd_min_limit: usize,
+    /// Maximum limit for AIMD algorithm. Makes sense only if `rate_limit_algorithm` is `Aimd`.
+    #[clap(long, default_value_t = 1500)]
+    aimd_max_limit: usize,
+    /// Increase AIMD increase by value in case of success. Makes sense only if `rate_limit_algorithm` is `Aimd`.
+    #[clap(long, default_value_t = 10)]
+    aimd_increase_by: usize,
+    /// Decrease AIMD decrease by value in case of timout/429. Makes sense only if `rate_limit_algorithm` is `Aimd`.
+    #[clap(long, default_value_t = 0.9)]
+    aimd_decrease_factor: f32,
+    /// A threshold below which the limit won't be increased. Makes sense only if `rate_limit_algorithm` is `Aimd`.
+    #[clap(long, default_value_t = 0.8)]
+    aimd_min_utilisation_threshold: f32,
```
can we extract these into a new struct and use `#[arg(flatten)]` https://docs.rs/clap/latest/clap/_derive/index.html#command-attributes

khanova replied:
Done.

khanova commented on proxy/src/rate_limiter/limit_algorithm.rs:
```diff
@@ -0,0 +1,59 @@
+    pub aimd_max_limit: usize,
+    pub aimd_increase_by: usize,
+    pub aimd_decrease_factor: f32,
+    pub aimd_min_utilisation_threshold: f32,
```
Done.
